server/plugins/DelayUGens.cpp

```diff
@@ -1600,6 +1609,37 @@ void RecordBuf_next_10(RecordBuf* unit, int inNumSamples) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+void BufIn_Ctor(BufIn* unit) {
+    unit->m_fbufnum = -1e9f;
+    unit->m_failedBufNum = -1e9f;
+
+    SETCALC(BufIn_next);
+    BufIn_next(unit, 1);
+}
+
+void BufIn_next(BufIn* unit, int inNumSamples) {
+    GET_BUF_SHARED
+    uint32 numOutputs = unit->mNumOutputs;
+    uint32 offset = static_cast<uint32>(ZIN0(1));
+
+    if (!bufData) {
+        if (unit->mWorld->mVerbosity > -1 && !unit->mDone && (unit->m_failedBufNum != fbufnum)) {
+            Print("Buffer UGen: no buffer data\n");
+            unit->m_failedBufNum = fbufnum;
+        }
+        ClearUnitOutputs(unit, inNumSamples);
+        return;
+    }
+
+    // we could wrap around buf samples, but no other ugen does it
+    if (offset < 0)
+        offset = 0;
+
+    for (size_t ch = offset; ch < numOutputs; ch++) {
+        OUT(ch)[0] = bufData[ch];
```

Comment:
what happens when the buffer has not enough samples?

Reply:
the ugen outputs zero. We could also wrap, it would make sense with other buf reading ugens, but it would make it more complicated and I don't know if there's a use case for it